File: NeuralNetwork/8puzzle.py
import heapq
# Manhattan distance heuristic
def manhattan_distance(state, goal_state):
    distance = 0
    for i in range(3):
        for j in range(3):
            if state[i][j] != -99:
                value = state[i][j]
                goal_i, goal_j = divmod(goal_state.index(value) if value in goal_state else 0, 3)
                distance += abs(i - goal_i) + abs(j - goal_j)
    return distance

# A* algorithm
def astar(initial_state, goal_state):
    priority_queue = [(manhattan_distance(initial_state, goal_state), 0, initial_state)]
    heapq.heapify(priority_queue)
    visited = set()

    while priority_queue:
        _, cost, current_state = heapq.heappop(priority_queue)

        print("Step:", cost)
        for row in current_state:
            print(row)
        print()

        if current_state == goal_state:
            return current_state

        if tuple(map(tuple, current_state)) in visited:
            continue

        visited.add(tuple(map(tuple, current_state)))

        for i in range(3):
            for j in range(3):
                if current_state[i][j] == -99:
                    for d in 'UDLR':
                        next_state = [row[:] for row in current_state]

                        if d == 'U' and i > 0:
                            next_state[i][j], next_state[i - 1][j] = next_state[i - 1][j], next_state[i][j]
                        elif d == 'D' and i < 2:
                            next_state[i][j], next_state[i + 1][j] = next_state[i + 1][j], next_state[i][j]
                        elif d == 'L' and j > 0:
                            next_state[i][j], next_state[i][j - 1] = next_state[i][j - 1], next_state[i][j]
                        elif d == 'R' and j < 2:
                            next_state[i][j], next_state[i][j + 1] = next_state[i][j + 1], next_state[i][j]

                        if tuple(map(tuple, next_state)) not in visited:
                            priority = cost + 1 + manhattan_distance(next_state, goal_state)
                            heapq.heappush(priority_queue, (priority, cost + 1, next_state))


# Initial and goal states
initial_state = [[1, 2, 3], [5, 6, -99], [4, 7, 8]]
goal_state = [[1, 2, 3], [4, 5, 6], [7, 8, -99]]

# Run A* algorithm
solution_state = astar(initial_state, goal_state)

File: NeuralNetworkINC/8puzzle.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 3

// Structure to represent a puzzle state
typedef struct {
    int puzzle[N][N];
} PuzzleState;

// Node structure for the priority queue
typedef struct {
    int priority;
    int cost;
    PuzzleState state;
} Node;

// Function to calculate the Manhattan distance heuristic
int manhattan_distance(PuzzleState state, PuzzleState goal_state) {
    int distance = 0;
    int i, j;
    for ( i = 0; i < N; i++) {
        for (j = 0; j < N; j++) {
            if (state.puzzle[i][j] != -99) {
                int value = state.puzzle[i][j];
                int goal_i, goal_j;
                for (goal_i = 0; goal_i < N; goal_i++) {
                    for (goal_j = 0; goal_j < N; goal_j++) {
                        if (value == goal_state.puzzle[goal_i][goal_j])
                            break;
                    }
                    if (goal_j < N)
                        break;
                }
                distance += abs(i - goal_i) + abs(j - goal_j);
            }
        }
    }
    return distance;
}

// Function to swap two puzzle elements
void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// Function to perform A* algorithm
PuzzleState astar(PuzzleState initial_state, PuzzleState goal_state) {
    Node priority_queue[N * N * N * N];
    int i , j;
    int front = 0, rear = 0;

    priority_queue[rear].priority = manhattan_distance(initial_state, goal_state);
    priority_queue[rear].cost = 0;
    priority_queue[rear].state = initial_state;
    rear++;

    PuzzleState visited[N * N * N * N];
    int visitedCount = 0;

    while (front < rear) {
        Node current = priority_queue[front++];
        PuzzleState current_state = current.state;

        printf("Step: %d\n", current.cost);
        for ( i = 0; i < N; i++) {
            for ( j = 0; j < N; j++) {
                printf("%d ", current_state.puzzle[i][j]);
            }
            printf("\n");
        }
        printf("\n");

        if (memcmp(&current_state, &goal_state, sizeof(PuzzleState)) == 0) {
            return current_state;
        }
		int k;
        int visitedFlag = 0;
        for (k = 0; k < visitedCount; k++) {
            if (memcmp(&current_state, &visited[k], sizeof(PuzzleState)) == 0) {
                visitedFlag = 1;
                break;
            }
        }

        if (visitedFlag)
            continue;

        visited[visitedCount++] = current_state;
		
        for ( i = 0; i < N; i++) {
            for ( j = 0; j < N; j++) {
                if (current_state.puzzle[i][j] == -99) {
                    for (char d : "UDLR") {
                        PuzzleState next_state = current_state;

                        if ((d == 'U' && i > 0) || (d == 'D' && i < N - 1) ||
                            (d == 'L' && j > 0) || (d == 'R' && j < N - 1)) {
                            int new_i = i + (d == 'D') - (d == 'U');
                            int new_j = j + (d == 'R') - (d == 'L');
                            swap(&next_state.puzzle[i][j], &next_state.puzzle[new_i][new_j]);

                            int visitedFlag = 0;
                            for (int k = 0; k < visitedCount; k++) {
                                if (memcmp(&next_state, &visited[k], sizeof(PuzzleState)) == 0) {
                                    visitedFlag = 1;
                                    break;
                                }
                            }

                            if (!visitedFlag) {
                                priority_queue[rear].priority = current.cost + 1 + manhattan_distance(next_state, goal_state);
                                priority_queue[rear].cost = current.cost + 1;
                                priority_queue[rear].state = next_state;
                                rear++;
                            }
                        }
                    }
                }
            }
        }
    }

    // Return the original state if no solution is found
    return initial_state;
}

int main() {
    // Initial and goal states
    PuzzleState initial_state = {{1, 2, 3}, {5, 6, -99}, {4, 7, 8}};
    PuzzleState goal_state = {{1, 2, 3}, {4, 5, 6}, {7, 8, -99}};

    // Run A* algorithm
    PuzzleState solution_state = astar(initial_state, goal_state);

    return 0;
}

